### ISSUE TITLE
Smtp is crashing, when using pre-connected socket in SMTPConnection options

### ISSUE THREAD

**Issue by kurevo1996** [nodemailer/smtp-connection#39]
Hello. I wanted to use this module, because i  wanted support for proxy, i connect a socket to proxy, and then just pass it in the constructor as a { connection: socket } parameter. However, it doesn't work. I'm using https://www.npmjs.com/package/socks and it works with proxies perfectly, the connection to proxy is established and everything is good. However then, when i try to login in smtp, following happens, here is the log:

```
Using ip 217.69.139.160 for host: smtp.mail.ru
[2016-02-14 23:10:39] INFO: [RcJrKOpceY] Secure connection established to 84.17.0.17:1080
[2016-02-14 23:10:39] DEBUG: [RcJrKOpceY] S: 220 smtp22.mail.ru ESMTP ready
[2016-02-14 23:10:39] DEBUG: [RcJrKOpceY] C: EHLO [127.0.0.1]
[2016-02-14 23:10:39] DEBUG: [RcJrKOpceY] S: 220 smtp22.mail.ru ESMTP ready
smtp established
[2016-02-14 23:10:39] DEBUG: [RcJrKOpceY] C: AUTH PLAIN AHNhZGc0MzYzNDZAbWFpbC5ydQBYTFNGb2cwc2V0ODIzNjBmYXNm
[2016-02-14 23:10:39] DEBUG: [RcJrKOpceY] S: 250-smtp22.mail.ru
250-SIZE 73400320
250-8BITMIME
250-PIPELINING
250 AUTH PLAIN LOGIN XOAUTH2
[2016-02-14 23:10:39] INFO: [RcJrKOpceY] User "sadg436346@mail.ru" authenticated
LOGGED IN
[2016-02-14 23:10:39] DEBUG: [RcJrKOpceY] S: 250-smtp22.mail.ru
250-SIZE 73400320
250-8BITMIME
250-PIPELINING
250 AUTH PLAIN LOGIN XOAUTH2
[2016-02-14 23:10:39] ERROR: [RcJrKOpceY] Unexpected Response: 250-smtp22.mail.ru
250-SIZE 73400320
250-8BITMIME
250-PIPELINING
250 AUTH PLAIN LOGIN XOAUTH2
smtp error: Error: Unexpected Response: 250-smtp22.mail.ru
250-SIZE 73400320
250-8BITMIME
250-PIPELINING
250 AUTH PLAIN LOGIN XOAUTH2
[2016-02-14 23:10:39] DEBUG: [RcJrKOpceY] Closing connection to the server using "end"
```

The 84.17.0.17:1080 is my proxy ip. 

I used the following code : 

```
function socksCallback(err, socket, info) {

		if(err){
			console.log(options.proxy.type.toString() + " : " + err);
			if(options.proxy.type >= 5) {
				console.log("bad proxy, 4/5 methods failed for " + options.target.host);
				return;
			}
			options.proxy.type++; // set socks5

			if(socket)
				socket.end();
			SocksClient.createConnection(options,socksCallback);

		}
		else {

			socket.resume();
			var connection = new SMTPConnection({connection : socket,connectionTimeout :20000, greetingTimeout: 20000,socketTimeout: 50000, debug: true,  logger:true,secure : true,requireTLS:false, tls: {
				rejectUnauthorized: false
			},
				authMethod : 'PLAIN'
			});
			connection.on("connect",function() {
				console.log("smtp established");

				connection.login({user : "myuser@mail.ru",pass :"mypass"},function(error,data){
					if(error)
						console.log(error);
					console.log("LOGGED IN");
				});

			});
			connection.on("error",function(err) {
				console.log("smtp error: " + err);


			});
			connection.connect();
			//socket.end();

		}

};
```

If i do it without passing proxy, everything works fine. Also if do it without pre-connected proxy socket, the server recognize TLS certificate names and i don't need to pass "rejectUnauthorized: false". Then what i did is i just tried to create regular net.Socket(), connect it and pass to the constructor like so:

```
var client = new net.Socket();
	client.connect(465, 'smtp.mail.ru', function() {
		console.log('Connected');

		var connection = new SMTPConnection({connection:client ,connectionTimeout :20000, greetingTimeout: 20000,socketTimeout: 50000, debug: true,  logger:true,secure : true,requireTLS:true, tls: {
			rejectUnauthorized: false
		},
			authMethod : 'PLAIN'
		});
		connection.on("connect",function() {
			console.log("smtp established");

			connection.login({user : "mymail@mail.ru",pass :"mypass"},function(error,data){
				if(error)
					console.log(error);
				console.log("LOGGED IN");

			});

		});
		connection.on("error",function(err) {
			console.log("smtp error: " + err);


		});
		connection.connect();
```

And i got almost same result, just with regular pre-connected non-proxy socket!!! 

```
[2016-02-14 23:07:03] INFO: [s0ahbEODdiU] Secure connection established to 217.69.139.160:465
[2016-02-14 23:07:03] DEBUG: [s0ahbEODdiU] S: 220 smtp44.i.mail.ru ESMTP ready
[2016-02-14 23:07:03] DEBUG: [s0ahbEODdiU] C: EHLO [127.0.0.1]
[2016-02-14 23:07:03] DEBUG: [s0ahbEODdiU] S: 220 smtp44.i.mail.ru ESMTP ready
smtp established
[2016-02-14 23:07:03] DEBUG: [s0ahbEODdiU] C: AUTH PLAIN AHNhZGc0MzYzNDZAbWFpbC5ydQBYTFNGb2cwc2V0ODIzNjBmYXNm
[2016-02-14 23:07:03] DEBUG: [s0ahbEODdiU] S: 250-smtp44.i.mail.ru
250-SIZE 73400320
250-8BITMIME
250-PIPELINING
250 AUTH PLAIN LOGIN XOAUTH2
[2016-02-14 23:07:03] INFO: [s0ahbEODdiU] User "sadg436346@mail.ru" authenticated
LOGGED IN
[2016-02-14 23:07:03] DEBUG: [s0ahbEODdiU] S: 250-smtp44.i.mail.ru
250-SIZE 73400320
250-8BITMIME
250-PIPELINING
250 AUTH PLAIN LOGIN XOAUTH2
[2016-02-14 23:07:03] ERROR: [s0ahbEODdiU] Unexpected Response: 250-smtp44.i.mail.ru
250-SIZE 73400320
250-8BITMIME
250-PIPELINING
250 AUTH PLAIN LOGIN XOAUTH2
smtp error: Error: Unexpected Response: 250-smtp44.i.mail.ru
250-SIZE 73400320
250-8BITMIME
250-PIPELINING
250 AUTH PLAIN LOGIN XOAUTH2
[2016-02-14 23:07:03] DEBUG: [s0ahbEODdiU] Closing connection to the server using "end"
```

Something is broken in the module, while working with pre-connected socket, i guess it's tls / ssl problem, even if i manually create normal socket and pass it in connection parameter, without "rejectUnauthorized: false" i get the error about my certificate. It is probably the case of this error **"Error: Unexpected Response: 250-smtp44.i.mail.ru"** Please help and give advice i can use pre-connected sockets corre

**Comment by kurevo1996** [comment]
I just tried something again, i looked up `._upgradeConnection` method and i found that there is a host used there in the 

```
    var opts = {
        socket: this._socket,
        host: this.options.host
    };

   this._socket = tls.connect(opts, function () {
```

So i added the ` host : 'smtp.mail.ru' ' in SMTPConnection constructor. But nothing has changed, i still got the `Error: Unexpected Response: 250-smtp44.i.mail.ru` in both cases with ordinary pre-connected socket, and with socks proxy connected socket.

**Comment by andris9** [comment]
Thanks, I'll try to check out what happens there. From the provided logs it seems that data events from the server are fired twice, so probably it's something to do how handlers for the events are added

**Comment by andris9** [comment]
There was a bug in the sequence of setting up event handlers as I suspected. I released the patched version as v2.2.5. Here's the script I used to verify the behavior: https://gist.github.com/andris9/679812feb3539a6b368e

btw. you do not need to set `rejectUnauthorized: false` if you have destination `host` and `port`, see the provided example how I connected to the server.
